### PR TITLE
Date & Age handling

### PR DESCRIPTION
**Pull request by benaadams**
Date & Age handling from #6

/cc @JunTaoLuo

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/IResponseCachingOptions.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.ResponseCaching
+{
+    interface IResponseCachingOptions
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/IResponseCachingOptions.cs]: Unused?

Class

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 2]
```diff
@@ -82,14 +98,24 @@ private string CreateCacheKey()
...
+                var age = Math.Max((DateTimeOffset.UtcNow - cacheEntry.Created).TotalSeconds, 0.0);
+                var ageString = (age > int.MaxValue ? int.MaxValue : (int)age).ToString(CultureInfo.InvariantCulture);
+                response.Headers[IsProxied ? "Age" : "X-Cache-Age"] = ageString;
 
-                // Copy the cached response body
-                var body = cacheEntry.Body;
-                if (body.Length > 0)
+                if (_requestType == RequestType.HeadersOnly)
+                {
+                    response.Headers["Content-Length"] = "0";
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Content-Length should be correct size for HEAD?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 3]
```diff
@@ -82,14 +98,24 @@ private string CreateCacheKey()
...
+                }
+                else
                 {
-                    await response.Body.WriteAsync(body, 0, body.Length);
+                    // Copy the cached response body
+                    var body = cacheEntry.Body;
+                    response.Headers["Content-Length"] = body.Length.ToString(CultureInfo.InvariantCulture);
+                    if (body.Length > 0)
+                    {
+                        await response.Body.WriteAsync(body, 0, body.Length);
```
Comment (vazzang) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Should this call have a `ConfigureAwait(false)` on the end of the await?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 4]
```diff
@@ -82,14 +98,24 @@ private string CreateCacheKey()
                     response.Headers[pair.Key] = pair.Value;
                 }
 
-                // TODO: Update cache headers (Age)
-                response.Headers["Served_From_Cache"] = DateTime.Now.ToString();
+                // TODO: Allow setting proxied _isProxied 
+                var age = Math.Max((DateTimeOffset.UtcNow - cacheEntry.Created).TotalSeconds, 0.0);
+                var ageString = (age > int.MaxValue ? int.MaxValue : (int)age).ToString(CultureInfo.InvariantCulture);
+                response.Headers[IsProxied ? "Age" : "X-Cache-Age"] = ageString;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: What is `IsProxied` for? What behaviour changes should this configure? Also, We should always be setting the `Age` header.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 5]
```diff
@@ -118,15 +144,37 @@ internal bool OnResponseStarting()
...
+            // Don't cache errors? 404 etc
+            if (CacheResponse && HttpContext.Response.StatusCode == 200)
             {
                 // Store the buffer to cache
-                var cacheEntry = new ResponseCachingEntry();
-                cacheEntry.StatusCode = HttpContext.Response.StatusCode;
-                foreach (var pair in HttpContext.Response.Headers)
+                var cacheEntry = new ResponseCachingEntry()
+                {
+                    Created = DateTimeOffset.UtcNow,
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: This should be set to the `Date` header from the response? That header contains the information on when the response was actually generated.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 6]
```diff
@@ -118,15 +144,37 @@ internal bool OnResponseStarting()
...
+                    StatusCode = HttpContext.Response.StatusCode
+                };
+
+                var headers = HttpContext.Response.Headers;
+                var count = headers.Count
+                    - (headers.ContainsKey("Date") ? 1 : 0)
+                    - (headers.ContainsKey("Content-Length") ? 1 : 0)
+                    - (headers.ContainsKey("Age") ? 1 : 0);
+                var cachedHeaders = new List<KeyValuePair<string, StringValues>>(count);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Hmm the values stored in here is just thrown away later on. Also, you can just set the values directly on `cacheEntry.Headers` now.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 7]
```diff
@@ -118,15 +144,37 @@ internal bool OnResponseStarting()
...
+                var age = 0;
+                foreach (var entry in headers)
                 {
-                    cacheEntry.Headers[pair.Key] = pair.Value;
+                    // Reduce create date by Age 
+                    if (entry.Key == "Age" && int.TryParse(entry.Value, out age) && age > 0)
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: I don't think we should rely on the `Age` header from upstream caches.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 8]
```diff
@@ -118,15 +144,37 @@ internal bool OnResponseStarting()
...
+                    {
+                        cacheEntry.Created -= new TimeSpan(0, 0, age);
+                    }
+                    // Don't copy Date header or Content-Length
+                    else if (entry.Key != "Date" && entry.Key != "Content-Length")
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: I think we should use the original date as provided by the origin. @Tratcher?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 9]
```diff
@@ -31,12 +35,24 @@ public ResponseCachingContext(HttpContext httpContext, IMemoryCache cache)
...
         {
             // Verify the method
             // TODO: What other methods should be supported?
-            if (!string.Equals("GET", HttpContext.Request.Method, StringComparison.OrdinalIgnoreCase))
+            if (string.Equals("GET", HttpContext.Request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                _requestType = RequestType.FullReponse;
+            }
+            else if (string.Equals("HEAD", HttpContext.Request.Method, StringComparison.OrdinalIgnoreCase) ||
+                     string.Equals("OPTIONS", HttpContext.Request.Method, StringComparison.OrdinalIgnoreCase))
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: RFC says "Responses to the OPTIONS method are not cacheable."

**Comment by JunTaoLuo** [comment]
This is a good start and I have this change rolled in with some more changes I made to implement a basic HTTP cache for which I will open a PR soon (I'm keeping your contribution in the commit history). There are still a few implementation details that I would like clarification though.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: It changes what Age means; e.g. if you are using ResponseCaching with https://github.com/aspnet/Proxy haven't quite worked out how they would communicate though

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 5 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Is when the cache entry is created?

Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: UtcNow at this point more accurately represents when the response was received by the cache, not when it was originally generated. The `Date` header is also what the RFC indicate to use for age calculations.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 4 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: From the RFC, we should always set `Age`. We might want to set the `X-Cache` headers but even that would be pending further discussion @Tratcher?

**Comment by JunTaoLuo** [comment]
@benaadams thanks for the contribution! Your changes have been merged in https://github.com/aspnet/ResponseCaching/commit/7d716d20072b50ea7c6f9e56cc31ed204de1cd85

**Pull request closed without merging**